File: ruby/day2/part1.rb
# frozen_string_literal: true

def parse_line(line)
  line.split(';').map do |game|
    game.split(',').to_h do |bag|
      count, colour = bag.strip.split
      [colour.to_sym, count.to_i]
    end
  end
end

def parse_game(game)
  bag = {
    red: 12,
    green: 13,
    blue: 14
  }

  game.each do |turn|
    reds = turn[:red] || 0
    blues = turn[:blue] || 0
    greens = turn[:green] || 0
    return false if reds > bag[:red] || blues > bag[:blue] || greens > bag[:green]
  end

  true
end

def part1(input)
  input.split("\n").reduce(0) do |result, line|
    game_id, game = line.match(/Game (\d+): (.*)/).captures
    turns = parse_line(game)
    if parse_game(turns)
      result + game_id.to_i
    else
      result
    end
  end
end

if __FILE__ == $PROGRAM_NAME
  input = File.read('input.txt')
  p part1(input)
end

File: elixir/day2/lib/part1.ex
defmodule Part1 do
  def parse_line(line) do
    line
    |> String.split(";")
    |> Enum.map(fn game ->
      game
      |> String.split(",")
      |> Enum.reduce(%{}, fn bag, acc ->
        [count, colour] = String.split(String.trim(bag))
        Map.put(acc, String.to_atom(colour), String.to_integer(count))
      end)
    end)
  end

  def parse_game(game) do
    bag = %{
      red: 12,
      green: 13,
      blue: 14
    }

    game
    |> Enum.all?(fn turn ->
      reds = turn[:red] || 0
      blues = turn[:blue] || 0
      greens = turn[:green] || 0

      reds <= bag[:red] && blues <= bag[:blue] && greens <= bag[:green]
    end)
  end

  def parse_game_id(game_id) do
    case Integer.parse(game_id) do
      {i, _} -> i
      _ -> 0
    end
  end

  def process_line(line, acc) do
    [_, game_id, game] = Regex.run(~r/Game (\d+): (.*)/, line)
    turns = parse_line(game)

    if parse_game(turns) do
      acc + parse_game_id(game_id)
    else
      acc
    end
  end

  def part1(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.reduce(0, &process_line/2)
  end
end

File.read!("input.txt")
|> Part1.part1()
|> IO.inspect(label: "part1")
